### PR TITLE
[CLA] Upgrade small strain isotropic damage CL

### PR DESCRIPTION
**Pull request by marandra**
**Description**
In this PR we are upgrading the hardening model of the smallstrainisotropicdamage CLs. Changes include the update of the current bilinear model to a multilinear one, and addition of a exponential hardening option. Test were updated, unused variables removed.

**Changelog**
- Update hardening models: Exponential and multilinear hardening
- Remove unused variable
- Added new variables
- Updated CLs tests.
- Minor rename of the variables, as discussed in #8635 

closes #8635

### CODE REVIEW

File: applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp [hunk 1]
```diff
@@ -516,7 +518,6 @@ void KratosConstitutiveLawsApplication::Register()
     KRATOS_REGISTER_VARIABLE(ACCUMULATED_PLASTIC_STRAIN)
     KRATOS_REGISTER_VARIABLE(BACK_STRESS_VECTOR)
     KRATOS_REGISTER_VARIABLE(BACK_STRESS_TENSOR)
-    KRATOS_REGISTER_VARIABLE(HARDENING_MODULI_VECTOR)
```
Comment (loumalouomega) [inline applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp:519]: Why removing this?

**Review by loumalouomega — CHANGES_REQUESTED** [review]
Comment

File: applications/ConstitutiveLawsApplication/custom_constitutive/small_strain_isotropic_damage_3d.cpp [hunk 2]
```diff
@@ -172,38 +173,42 @@ void SmallStrainIsotropicDamage3D::CalculateStressResponse(
...
         {
             // INELASTIC
-            strain_variable = strain_norm;
-            const double stress_variable = EvaluateHardeningLaw(strain_variable, r_material_properties);
-            const double damage_variable = 1. - stress_variable / strain_variable;
-            const double hardening_modulus = EvaluateHardeningModulus(strain_variable, r_material_properties);
-            const double damage_rate = (stress_variable - hardening_modulus * strain_variable)
-                                       / (strain_variable * strain_variable * strain_variable);
-            r_constitutive_matrix *= (1. - damage_variable);
+            r = strain_norm;
```
Comment (AlejandroCornejo) [inline applications/ConstitutiveLawsApplication/custom_constitutive/small_strain_isotropic_damage_3d.cpp:199]: I know why you did this, no comments hehe

**Review by AlejandroCornejo — DISMISSED** [review]

File: applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_3d.cpp [hunk 3]
```diff
@@ -96,201 +102,141 @@ KRATOS_TEST_CASE_IN_SUITE(_ConstitutiveLaw_SmallStrainIsotropicDamage3D, KratosC
...
-        }
-    }
+    // Check stress
+    for (std::size_t comp = 0; comp < 6; ++comp)
+        KRATOS_CHECK_NEAR(stress_vector(comp), ref_stress(comp), tolerance);
 
+    // Check constitutive tensor
+    for (std::size_t i = 0; i < 6; ++i)
+        for (std::size_t j = 0; j < 6; ++j)
+            KRATOS_CHECK_NEAR(const_matrix(i, j), ref_C(j, i), tolerance);
```
Comment (philbucher) [inline applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_3d.cpp:166]: better use KRATOS_CHECK_VECTOR/MATRIX if possible

File: applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp (hunk 1 above)
Comment (marandra) [reply applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp:519]: We originally added it for this CLs, but now we changed some of the parameters and this one no longer applies. I checked and it's not in use in Kratos, and tests run normally.

File: applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_3d.cpp (hunk 3 above)
Comment (marandra) [reply applications/ConstitutiveLawsApplication/tests/cpp_tests/test_small_strain_isotropic_damage_3d.cpp:166]: Didn´t know those. I'm planning a followup PR, updating these same tests for testing behavior under compression, and moving 2D tests from python to cpp. I'll fix the KRATOS_CHECK_MATRIX issue there if you don´t mind.

File: applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp (hunk 1 above)
Comment (marandra) [reply applications/ConstitutiveLawsApplication/constitutive_laws_application.cpp:519]: Is it OK or should I leave it?

**Comment by marandra** [comment]
ping @loumalouomega

**Review by loumalouomega — APPROVED** [review]

**Comment by marandra** [comment]
Thanks!

**Pull request merged**